### ISSUE TITLE
DATA and DEROUTPUT do not have compatible formats in vl_nnconv

### ISSUE THREAD

**Issue by AkaiZhen** [AkaiZhen/CNN_Trianing#1]
Hi, I run into the following issue when I use the cpu version on the new created network. The array is already single. MATLAB version is R2014b. Operating System is Win7 64bits. Any clue what that error is?

Error using vl_nnconv
DATA and DEROUTPUT do not have compatible formats.

This is my source code:

Network definition:
%---------------------------network define--------------------------------
net.layers{1} = struct('name', 'conv1', ...
    'type', 'conv', ...
    'weights', {{randn(5, 5, 3, 24, 'single'), randn(24, 1, 'single')}}, ...
    'pad', 0, ...
    'stride', 2, ...
    'dilate', 1);
net.layers{2} = struct('name', 'conv2', ...
    'type', 'conv', ...
    'weights', {{randn(5, 5, 24, 36, 'single'), randn(36, 1, 'single')}}, ...
    'pad', 0, ...
    'stride', 2, ...
    'dilate', 1);
net.layers{3} = struct('name', 'conv3', ...
    'type', 'conv', ...
    'weights', {{randn(5, 5, 36, 48, 'single'), randn(48, 1, 'single')}}, ...
    'pad', 0, ...
    'stride', 2, ...
    'dilate', 1);
net.layers{4} = struct('name', 'conv4', ...
    'type', 'conv', ...
    'weights', {{randn(3, 3, 48, 64, 'single'), randn(64, 1, 'single')}}, ...
    'pad', 0, ...
    'stride', 1, ...
    'dilate', 1);
net.layers{5} = struct('name', 'conv5', ...
    'type', 'conv', ...
    'weights', {{randn(3, 3, 64, 64, 'single'), randn(64, 1, 'single')}}, ...
    'pad', 0, ...
    'stride', 1, ...
    'dilate', 1);
%----------------------------end of define--------------------------------

Training function:
function [net, stats] = dave2_train(net, varargin)

opts.train.batchSize = 200 ;
opts.train.numEpochs = 10 ;
opts.train.continue = true ;
opts.train.gpus = [] ;
opts.train.learningRate = 0.01 ;
opts.train.expDir = [vl_rootnn '/data/dave2'] ;
opts.dataDir = [vl_rootnn '/data/dave2-dataset'] ;
[opts, varargin] = vl_argparse(opts, varargin) ;

opts.imdbPath = [opts.train.expDir '/imdb.mat'] ;
opts = vl_argparse(opts, varargin) ;

% --------------------------------------------------------------------
%                             Prepare data
% --------------------------------------------------------------------

% Create image database (imdb struct). It can be cached to a file for speed
if exist(opts.imdbPath, 'file')
  disp('Reloading image database...')
  imdb = load(opts.imdbPath) ;
else
  disp('Creating image database...')
  imdb = getImdb(opts.dataDir) ;
  mkdir(fileparts(opts.imdbPath)) ;
  save(opts.imdbPath, '-struct', 'imdb') ;
end

% --------------------------------------------------------------------
%                              Train
% --------------------------------------------------------------------

use_gpu = ~isempty(opts.train.gpus) ;

% Start training
[net, stats] = cnn_train(net, imdb, @(imdb, batch) getBatch(imdb, batch, use_gpu), 'train', find(imdb.set == 1), 'val', find(imdb.set == 2), opts.train) ;

% Visualize the learned filters
%figure(1) ; vl_tshow(net.layers{1}.weights{1}) ; title('Conv1 filters') ;
%figure(2) ; vl_tshow(net.layers{3}.weights{1}) ; title('Conv2 filters') ;
%figure(3) ; vl_tshow(net.layers{5}.weights{1}) ; title('Conv3 filters') ;

% --------------------------------------------------------------------
function [images, labels] = getBatch(imdb, batch, use_gpu)
% --------------------------------------------------------------------
% This is where we return a given set of images (and their labels) from
% our imdb structure.
% If the dataset was too large to fit in memory, getBatch could load images
% from disk instead (with indexes given in 'batch').

images = imdb.images(:,:,:,batch) ;
labels = imdb.labels(batch) ;

if use_gpu
  images = gpuArray(images) ;
end

% --------------------------------------------------------------------
function imdb = getImdb(dataDir)
% --------------------------------------------------------------------
% Initialize the imdb structure (image database).
% Note the fields are arbitrary: only your getBatch needs to understand it.
% The field imdb.set is used to distinguish between the training and
% validation sets, and is only used in the above call to cnn_train.

% The sets, and number of samples per label in each set
sets = {'train', 'val'} ;
numSamples = [250, 10] ;

% Preallocate memory
totalSamples = 1040 ;  % 4 * 250 + 4 * 10
images = zeros(66, 640, 3, totalSamples, 'single') ;
labels = zeros(totalSamples, 1) ;
set = ones(totalSamples, 1) ;

% Read all samples
sample = 1 ;
for s = 1:2  % Iterate sets
    for label = 1:4  % Iterate labels
        for i = 1:numSamples(s)  % Iterate samples
            % Read image
            im = imread(sprintf('%s/%s/%i/%04i.jpg', dataDir, sets{s}, label, i)) ;
            
            % Store it, along with label and train/val set information
            images(:,:,:,sample) = single(im) ;
            labels(sample) = label ;
            set(sample) = s ;
            sample = sample + 1 ;
        end
    end
end

% Store results in the imdb struct
imdb.images = images ;
imdb.labels = labels ;
imdb.set = set ;
